In [1]:
import pandas as pd
import numpy as np

In [2]:
df=pd.read_csv("Downloads/dirty_customer_dataset_100k.csv")

In [3]:
df.shape

(101500, 18)

In [4]:
df.isnull().sum()

CustomerID                   0
Full_Name                    0
Age                       2009
Gender                       0
Country                      0
City                         0
Email                     2004
Phone                     1011
Signup_Date                  0
Last_Purchase_Date         509
Total_Purchase_Amount        0
Num_Orders                  87
Product_Category             0
Rating                   17916
Loyalty_Tier                 0
Annual_Income             1526
Discount_Used                0
Churn_Status                 0
dtype: int64

# Remove Duplicates


In [5]:
df.duplicated().sum()

1439

In [6]:
df=df.drop_duplicates()

In [8]:
df.shape

(100061, 18)

## Split first and last name

In [15]:
df = df.copy()
df[['First_name', 'Last_name']] = df['Full_Name'].str.split(n=1, expand=True)

In [17]:
df.drop(columns='Full_Name',inplace=True)

In [25]:
df=df[['CustomerID','First_name',  'Last_name','Age', 'Gender', 'Country', 'City', 'Email', 'Phone',
       'Signup_Date', 'Last_Purchase_Date', 'Total_Purchase_Amount',
       'Num_Orders', 'Product_Category', 'Rating', 'Loyalty_Tier',
       'Annual_Income', 'Discount_Used', 'Churn_Status' ]]

In [27]:
df["Age"]=df["Age"].astype("Int64")

In [29]:
df.dtypes

CustomerID                object
First_name                object
Last_name                 object
Age                        Int64
Gender                    object
Country                   object
City                      object
Email                     object
Phone                     object
Signup_Date               object
Last_Purchase_Date        object
Total_Purchase_Amount     object
Num_Orders                object
Product_Category          object
Rating                   float64
Loyalty_Tier              object
Annual_Income            float64
Discount_Used             object
Churn_Status              object
dtype: object

## Fix Age 

In [31]:
df[(df["Age"]<18) | (df["Age"]>80)].shape[0]

500

In [33]:
df["Age"] = pd.to_numeric(df["Age"], errors="coerce")

In [35]:
df.loc[(df["Age"] < 18) | (df["Age"] > 80), "Age"] = np.nan

In [37]:
df["Age"].isnull().sum()

2489

## Clean Gender

In [39]:
df.loc[(df["Gender"].str.lower()=='female') | (df["Gender"].str.lower()=='f') ,"Gender"]='F'
df.loc[(df["Gender"].str.lower()=='male') | (df["Gender"].str.lower()=='m') ,"Gender"]='M'

In [41]:
df[~df["Gender"].isin(['F','M'])]

,CustomerID,First_name,Last_name,Age,Gender,Country,City,Email,Phone,Signup_Date,Last_Purchase_Date,Total_Purchase_Amount,Num_Orders,Product_Category,Rating,Loyalty_Tier,Annual_Income,Discount_Used,Churn_Status


## Clean Country

In [43]:
df["Country"] = df["Country"].str.strip()

In [45]:
df.loc[(df["Country"].str.lower()=="uk") |(df["Country"].str.lower()=="united kingdom"),"Country"]='UK'
df.loc[(df["Country"].str.lower()=="pk") |(df["Country"].str.lower()=="pakistan"),"Country"]='PK'
df.loc[(df["Country"].str.lower()=="usa") | (df["Country"].str.lower()=="united states") ,"Country"]='USA'

In [47]:
df[~df["Country"].isin(['USA','UK','PK','GB'])]

,CustomerID,First_name,Last_name,Age,Gender,Country,City,Email,Phone,Signup_Date,Last_Purchase_Date,Total_Purchase_Amount,Num_Orders,Product_Category,Rating,Loyalty_Tier,Annual_Income,Discount_Used,Churn_Status


## Fix Invaid Emails

In [49]:
df["Email"]=df["Email"].str.strip()

In [51]:
df.loc[df["Email"].str.contains("@@", na=False),"Email"]=np.nan
df.loc[~df["Email"].str.contains("@", na=False),"Email"]=np.nan

In [53]:
df["Email"].isnull().sum()

2204

## Fix Invalid Phone Numbers

In [55]:
df["Phone"]=df["Phone"].str.strip()

In [57]:
df["Phone"] = df["Phone"].str.replace(r"[-./\\]", "", regex=True)
df["Phone"] = df["Phone"].str.split("x").str[0]

In [59]:
df["Phone"]=df["Phone"].str.strip()

In [61]:
df.loc[df["Phone"].str.contains(r"[a-zA-Z]", na=False), "Phone"] = np.nan

# Fix invalid Future Signup Dates

In [63]:
df["Signup_Date"]=pd.to_datetime(df["Signup_Date"])

In [65]:
df["Last_Purchase_Date"]=pd.to_datetime(df["Last_Purchase_Date"])

## Clean Amount Values

In [67]:
df["Total_Purchase_Amount"]=df["Total_Purchase_Amount"].str.replace(r"[Rs.£$USD]","", regex=True)
df["Total_Purchase_Amount"]=df["Total_Purchase_Amount"].astype("float")
df.loc[df["Total_Purchase_Amount"]<0,"Total_Purchase_Amount"]=np.nan

In [69]:
df["Num_Orders"]=df["Num_Orders"].str.replace(r"[^\d]","", regex=True)
df["Num_Orders"] = df["Num_Orders"].replace("", np.nan)
df["Num_Orders"]=df["Num_Orders"].astype("float")

In [71]:
df.loc[df["Num_Orders"]<0,"Num_Orders"]=np.nan

## Fix/Encode All Categorical Values

In [73]:
df["Product_Category"].unique()

array(['Food', 'CLOTHING', 'Clothing', 'Beauty', 'electronics', 'Books',
       'BOOKS', 'Electronics', 'food'], dtype=object)

In [75]:
df["Product_Category"] = df["Product_Category"].str.strip().str.title()

In [77]:
df["Product_Category"].unique()

array(['Food', 'Clothing', 'Beauty', 'Electronics', 'Books'], dtype=object)

In [79]:
df["Rating"]=df["Rating"].astype(float)

In [81]:
df["Loyalty_Tier"] = df["Loyalty_Tier"].str.strip().str.title()

In [83]:
df["Loyalty_Tier"].unique()

array(['Platinum', 'Bronze', 'Silver', 'Gold'], dtype=object)

In [85]:
df["Annual_Income"]=df["Annual_Income"].astype("float")

In [87]:
df.loc[df["Annual_Income"]<0,"Annual_Income"]=np.nan

In [89]:
df["Discount_Used"]=df["Discount_Used"].str.strip().str.title()

In [91]:
df["Discount_Used"]=df["Discount_Used"].str.replace("1","Yes")
df["Discount_Used"]=df["Discount_Used"].str.replace("True","Yes")
df["Discount_Used"]=df["Discount_Used"].str.replace("0","No")
df["Discount_Used"]=df["Discount_Used"].str.replace("False","No")

In [93]:
df["Churn_Status"]=df["Churn_Status"].str.strip().str.title()

In [95]:
df["Churn_Status"]=df["Churn_Status"].str.replace("Churned","1")
df["Churn_Status"]=df["Churn_Status"].str.replace("Active","0")

In [97]:
today=pd.Timestamp.today()

In [98]:
df.loc[df["Signup_Date"] > today, "Signup_Date"] = pd.NaT

In [101]:
df.loc[df["Last_Purchase_Date"] > today, "Last_Purchase_Date"] = pd.NaT

In [103]:
df

,CustomerID,First_name,Last_name,Age,Gender,Country,City,Email,Phone,Signup_Date,Last_Purchase_Date,Total_Purchase_Amount,Num_Orders,Product_Category,Rating,Loyalty_Tier,Annual_Income,Discount_Used,Churn_Status
0,CUST094929,John,Brown II,39,F,UK,Christopherbury,pkirby@example.net,7087539740,2024-09-10 21:26:19.299793,2024-09-30 08:26:44.464044,495084.0,1050.0,Food,5.0,Platinum,22099.98,Yes,1
1,CUST062692,Paul,Lewis,22,F,USA,West Julian,susan24@example.net,4183783214,2022-10-05 13:45:54.351543,2023-02-19 09:01:28.452884,259197.0,330.0,Food,1.0,Bronze,50388.70,Yes,1
2,CUST049801,Amanda,Farmer,<NA>,M,USA,Bakerstad,kelly82@example.com,8369920214,2021-12-27 03:05:37.935379,2022-06-28 20:31:20.514805,7875.0,690.0,Clothing,1.0,Silver,36724.89,No,1
3,CUST016670,Christine,Brown,33,F,USA,Kimberlystad,anthony44@example.net,0015049772928,2020-01-01 05:18:52.607326,2020-10-31 13:05:46.359463,451114.0,770.0,Clothing,3.0,Silver,131100.03,Yes,1
4,CUST047156,Julie,Knight,57,M,PK,New Alexis,carterpreston@example.net,5394713338,2021-10-30 04:13:59.384393,2022-05-11 12:43:52.202322,75752.0,220.0,Beauty,3.0,Gold,116423.93,No,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
101494,CUST057043,Wendy,Henderson,29,M,USA,Rogersville,kingjacob@example.com,2274341620,2022-06-03 19:15:54.837548,2022-11-08 04:01:34.608946,145514.0,300.0,Beauty,NaN,Gold,156719.35,Yes,0
101496,CUST036009,Jesse,Cruz,65,M,PK,Alexishaven,marymalone@example.com,6462259986,2021-02-27 22:38:09.838898,2021-10-19 20:56:44.626046,246661.0,1340.0,Clothing,4.0,Platinum,171466.84,No,1
101497,CUST046266,April,Buchanan,20,M,PK,Peckville,tsullivan@example.net,0016544707223,2021-10-10 16:13:51.176311,2022-04-25 06:28:48.233282,162768.0,1030.0,Books,2.0,Platinum,61206.12,No,1
101498,CUST023588,Richard,Myers,32,F,USA,Susanshire,darrenwashington@example.com,7982834280,2020-05-31 19:06:43.600036,2021-03-06 22:31:52.267122,418013.0,430.0,Clothing,5.0,Bronze,172789.73,Yes,0


In [107]:
df.to_csv("Downloads/Cleaned_Dataset.csv", index=False)

In [113]:
df["Churn_Status"]

0         1
1         1
2         1
3         1
4         0
         ..
101494    0
101496    1
101497    1
101498    0
101499    1
Name: Churn_Status, Length: 100061, dtype: object

In [121]:
df=df.drop_duplicates()

In [123]:
df.duplicated().sum()

0

In [125]:
df["Churn_Status"].value_counts()

Churn_Status
0    57175
1    42881
Name: count, dtype: int64

In [139]:
df=df.copy()
df["Age"]=df["Age"].fillna(df["Age"].median())

In [160]:
df.shape

(100056, 19)

## Project Summary

This notebook documents a complete data cleaning workflow performed on a synthetic 
customer dataset of 101,500 rows and 18 columns, replicating real-world data quality 
issues commonly encountered in business environments.

### What Was Done

| Step | Action | Rows/Values Affected |
|---|---|---|
| Duplicate removal | Dropped exact duplicate records | 1,500 rows removed |
| Invalid ages | Replaced values outside 18–80 with NaN, imputed with median | 500+ values |
| Inconsistent categories | Standardized Gender, Country, Loyalty, Category, Discount, Churn | All string columns |
| Malformed emails | Replaced invalid formats with NaN | 600+ values |
| Currency symbols | Stripped $, £, Rs., USD using regex, converted to float | 350 values |
| Negative amounts | Replaced with NaN | 300 values |
| Wrong data types | Cleaned non-numeric strings in Num_Orders, converted to float | 400 values |
| Future dates | Replaced 2027 signup dates with NaT | 200 values |
| Feature engineering | Split Full_Name into First_name and Last_name | All rows |
| Label encoding | Encoded Churn_Status: Active = 0, Churned = 1 | All rows |

### Intentional Remaining Nulls

Rating nulls (17,634) were left as NaN deliberately. A missing rating means the 
customer did not rate, which is meaningful information. Imputing a rating value 
would misrepresent customer behaviour and introduce bias into any downstream analysis.

Email and phone nulls were also retained as NaN. Contact information cannot be 
estimated or imputed.

### Final Dataset

- **Rows:** 100,056  
- **Columns:** 19  
- **Status:** Clean and analysis-ready  

### Tools Used

Python 3 · pandas · NumPy · Jupyter Notebook

---
*Project by Areesha Kamil — [GitHub](https://github.com/AreeshaSolangi) · 
[LinkedIn](https://linkedin.com/in/areesha-kamil-5607b4265)*